In [3]:
# %matplotlib inline

Bit Error Rate (BER)

*  The Bit Error Rate (BER) is a measure of how many bits of data are incorrectly received compared to the total number of bits sent over a communication channel. It is expressed as a ratio or percentage.

*  The formula for calculating BER is:
  
*  BER = Number of Bit Errors / Total Number of Bits Sent


*  If you send 1,000 bits and 10 of those bits are received incorrectly, the BER would be:
*  BER = 10/1000 = 0.01 or 1%

Block Error Rate (BLER)
Definition:
The Block Error Rate (BLER) measures the number of blocks of data that are received incorrectly compared to the total number of blocks sent. A "block" typically refers to a group of bits that are transmitted together, often as part of a larger message.


- The formula for calculating BLER is:
 BLER = Number of Block Errors / Total Number of Blocks Sent



- If you send 100 blocks of data and 5 of those blocks are received incorrectly, the BLER would be:
BLER = 5/100 = 5%

In [1]:
import importlib
import AIsim_main2  # Ensure the module is imported

In [2]:
# Reload the module to reflect changes
importlib.reload(AIsim_main2)

from AIsim_main2 import sim_bersingle2, Transmitter

In [3]:
# Define the necessary variables before calling the function
dataset_folder = r"D:\Research\AIsensing\deeplearning\Database"
scenario = "O1_60"

NUM_UT = 1
NUM_BS = 1
NUM_UT_ANT = 1
NUM_BS_ANT = 16
NUM_BITS_PER_SYMBOL = 2
BATCH_SIZE = 128
showfigure = True
datapathbase = "data/"

eval_transceiver = Transmitter(
    channeldataset="deepmimo",
    channeltype="ofdm",
    scenario=scenario,
    dataset_folder=dataset_folder,
    direction="uplink",
    num_ut=NUM_UT,
    num_ut_ant=NUM_UT_ANT,
    num_bs=NUM_BS,
    num_bs_ant=NUM_BS_ANT,
    batch_size=BATCH_SIZE,
    fft_size=76,
    num_ofdm_symbols=14,
    num_bits_per_symbol=NUM_BITS_PER_SYMBOL,
    subcarrier_spacing=60e3,
    USE_LDPC=False,
    pilot_pattern="kronecker",
    guards=True,
    showfig=showfigure,
    savedata=True,
    outputpath=datapathbase,  # deepmimo_ofdm_
)
# # Now you can call the function
# print("running deepmimo dataset")
# bers, blers, BERs = sim_bersingle2(
#     channeldataset="deepmimo",
#     channeltype="ofdm",
#     NUM_BITS_PER_SYMBOL=2,
#     EBN0_DB_MIN=-5.0,
#     EBN0_DB_MAX=25.0,
#     BATCH_SIZE=128,
#     NUM_UT=1,
#     NUM_BS=1,
#     NUM_UT_ANT=1,
#     NUM_BS_ANT=16,
#     showfigure=True,  # or whatever value you want
#     datapathbase="data/",
#     dataset_folder=dataset_folder
# )

# # Add this line to display the figures
# import matplotlib.pyplot as plt

# plt.show()  # Ensure figures are displayed

Average energy per pilot symbol: 1.00
RG num_ofdm_symbols 14
RG ofdm_symbol_duration 1.7982456140350878e-05
512

Basestation 1

UE-BS Channels


Generating channels: 100%|██████████| 18100/18100 [00:00<00:00, 18165.37it/s]



BS-BS Channels


Generating channels: 100%|██████████| 1/1 [00:00<?, ?it/s]


dict_keys(['paths', 'LoS', 'location', 'distance', 'pathloss', 'channel'])
(18100, 3)
[242.42300415 297.17098999   2.        ]
1
dict_keys(['user', 'basestation', 'location'])
dict_keys(['paths', 'LoS', 'location', 'distance', 'pathloss', 'channel'])
18100
(18100, 1, 16, 10)
(18100, 1, 16, 10)
[[[-3.40252654e-06-1.42072656e-06j  1.81741018e-06+2.55279360e-06j
   -2.07117159e-06+1.63894288e-06j -1.15924456e-06-4.15754414e-07j
    1.61744200e-08+6.18089018e-07j -4.61281388e-07-3.27500629e-07j
    3.21317053e-08+1.42540202e-07j  2.89082109e-08-1.40979708e-07j
    7.96548427e-09+4.98924386e-08j  0.00000000e+00+0.00000000e+00j]
  [-2.97919746e-06-2.17256388e-06j  2.91234210e-06+1.15671912e-06j
   -2.39480664e-06+1.11390580e-06j -1.19963624e-06+2.78518655e-07j
   -4.32637762e-07-4.41724154e-07j  5.60680348e-07-7.53309948e-08j
    6.10191151e-08-1.32766047e-07j -1.08757888e-07+9.42479943e-08j
    1.72580816e-09-5.04948119e-08j  0.00000000e+00+0.00000000e+00j]
  [-2.39273436e-06-2.80543645e-06

Generating channels: 100%|██████████| 18100/18100 [00:01<00:00, 17217.61it/s]



BS-BS Channels


Generating channels: 100%|██████████| 1/1 [00:00<?, ?it/s]


dict_keys(['paths', 'LoS', 'location', 'distance', 'pathloss', 'channel'])
(18100, 3)
[242.42300415 297.17098999   2.        ]
1
dict_keys(['user', 'basestation', 'location'])
dict_keys(['paths', 'LoS', 'location', 'distance', 'pathloss', 'channel'])
18100
(18100, 1, 16, 10)
(18100, 1, 16, 10)
[[[-3.40252654e-06-1.42072656e-06j  1.81741018e-06+2.55279360e-06j
   -2.07117159e-06+1.63894288e-06j -1.15924456e-06-4.15754414e-07j
    1.61744200e-08+6.18089018e-07j -4.61281388e-07-3.27500629e-07j
    3.21317053e-08+1.42540202e-07j  2.89082109e-08-1.40979708e-07j
    7.96548427e-09+4.98924386e-08j  0.00000000e+00+0.00000000e+00j]
  [-2.97919746e-06-2.17256388e-06j  2.91234210e-06+1.15671912e-06j
   -2.39480664e-06+1.11390580e-06j -1.19963624e-06+2.78518655e-07j
   -4.32637762e-07-4.41724154e-07j  5.60680348e-07-7.53309948e-08j
    6.10191151e-08-1.32766047e-07j -1.08757888e-07+9.42479943e-08j
    1.72580816e-09-5.04948119e-08j  0.00000000e+00+0.00000000e+00j]
  [-2.39273436e-06-2.80543645e-06

# Transmitter method

## bit to be transferred

In [162]:
from deepMIMO5 import (

    BinarySource,
    calculate_BER,
)

In [163]:
binary_source = BinarySource()
NUM_STREAMS_PER_TX = eval_transceiver.num_streams_per_tx
k = eval_transceiver.k
b = binary_source(
        [BATCH_SIZE, 1, NUM_STREAMS_PER_TX, k]
    )  # [batch_size, num_tx, num_streams_per_tx, num_databits]

In [164]:
b.shape

(128, 1, 1, 1536)

In [165]:
# Now call the gettransmitSignal method
y, x_rg, x, b, no, h_b, tau_b, h_out = eval_transceiver.gettransmitSignal(
    b=b,  # You can pass your bits here if you have them
    ebno_db=15.0,  # Example Eb/No valueA
    perfect_csi=False,  # Example for perfect CSI
)

# Print the results
print("Received signal (y):", y)
print("Resource grid (x_rg):", x_rg)
print("Transmitted signal (x):", x)
print("Original bits (b):", b)
print("Noise power (no):", no)
print("Channel impulse response (h_b):", h_b)
print("Delay profile (tau_b):", tau_b)
print("Channel response (h_out):", h_out)

h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
Received signal (y): [[[[[ 0.15052705-3.18893902e-02j  0.10013501+3.07297260e-02j
      0.12948984-5.05431229e-03j ... -0.12063538-1.09146526e-02j
      0.07423247+6.17971085e-02j  0.00117465+9.30223987e-02j]
    [-0.0256925 -5.94639853e-02j  0.14839135+1.43046156e-02j
     -0.00521123+1.46622881e-01j ... -0.03897303+1.12219322e-02j
      0.23114121-7.24919513e-02j  0.14008693+7.79391900e-02j]
    [ 0.07629317-8.23451355e-02j  0.11678226+2.68992186e-02j
     -0.08859797-1.01385772e-01j ... -0.16358267-4.69600894e-02j
     -0.11477481-8.04143548e-02j -0.1112001 +2.08746359e-01j]
    ...
    [-0.0166862 +1.55625502e-02j -0.07958218-5.66568039e-03j
     -0.05902097+1.32118901e-02j ... -0.01943209+1.68602355e-02j
      0.02734549-5.55924326e-02j  0.04047745+1.01162400e-02j]
    [ 0.16184062-1.18963897e-01j  0.00081691+2.17346754e-02j
     -0.16259864+1.41763866e-01j ... -0.0696109 +1.24661557e-01j
     -0.05744087+1.14081

In [166]:
# Print the results
print("Received signal (y):", y.shape)
print("Resource grid (x_rg):", x_rg.shape)
print("Transmitted signal (x):", x.shape)
print("Original bits (b):", b.shape)
print("Noise power (no):", no)
print("Channel impulse response (h_b):", h_b.shape)
print("Delay profile (tau_b):", tau_b.shape)
print("Channel response (h_out):", h_out.shape)

Received signal (y): (128, 1, 1, 14, 76)
Resource grid (x_rg): (128, 1, 1, 14, 76)
Transmitted signal (x): (128, 1, 1, 768)
Original bits (b): (128, 1, 1, 1536)
Noise power (no): 0.015811387
Channel impulse response (h_b): (128, 1, 1, 1, 16, 10, 1)
Delay profile (tau_b): (128, 1, 1, 10)
Channel response (h_out): (128, 1, 1, 1, 16, 1, 76)


# Baseline Receiver

In [25]:
print(no)
x_hat, no_eff, h_hat, err_var, h_perfect, err_var_perfect = eval_transceiver.channelest_equ(
            y, no, h_b=h_b, tau_b=tau_b, h_out=h_out, perfect_csi=False
        )

# print types
# Print shapes
print("x_hat shape:", x_hat.shape)
print("no_eff shape:", no_eff.shape)
print("h_hat shape:", h_hat.shape)

      
# Print the results
# print("Estimated transmitted signal (x_hat):", x_hat)
print("Effective noise power (no_eff):", no_eff)
print("Estimated channel impulse response (h_hat):", h_hat)
print("Error variance (err_var):", err_var)
print("Perfect channel impulse response (h_perfect):", h_perfect)
print("Error variance for perfect CSI (err_var_perfect):", err_var_perfect)

0.015811387
Num of Pilots: 1
Num of Pilots: 1
h_out after remove nulled shape: (128, 1, 1, 1, 16, 1, 64)
x_hat shape: (128, 1, 1, 768)
no_eff shape: ()
h_hat shape: (128, 1, 1, 1, 1, 14, 64)
Effective noise power (no_eff): 0.13165171
Estimated channel impulse response (h_hat): tf.Tensor(
[[[[[[[ 3.393912  -1.6277184j   3.179735  -1.8958029j
        2.8360512 -2.1245131j  ...  3.6501496 -0.24027613j
        3.5932324 -0.6036585j   3.5210452 -0.9691628j ]
      [ 3.393912  -1.6277184j   3.179735  -1.8958029j
        2.8360512 -2.1245131j  ...  3.6501496 -0.24027613j
        3.5932324 -0.6036585j   3.5210452 -0.9691628j ]
      [ 3.393912  -1.6277184j   3.179735  -1.8958029j
        2.8360512 -2.1245131j  ...  3.6501496 -0.24027613j
        3.5932324 -0.6036585j   3.5210452 -0.9691628j ]
      ...
      [ 3.3905582 -1.642009j    3.1866398 -1.8298731j
        3.0492642 -2.154175j   ...  3.607661  -0.30493242j
        3.6949346 -0.661312j    3.5552833 -1.0135303j ]
      [ 3.3905582 -1.6420

In [27]:
b_hat, llr_est = eval_transceiver.demapper_decision(x_hat=x_hat, no_eff=no_eff)

BER = calculate_BER(b, b_hat)
print("BER Value:", BER)

BER Value: 0.009562174479166666


# Neural receiver v1 ber high

# Neural receiver v2

In [88]:
# import torch
# import torch.nn as nn
# import numpy as np


# class NeuralReceiver(nn.Module):
#     def __init__(self):
#         """
#         Neural Receiver that processes received signals to remove noise and channel effects.
#         """
#         super(NeuralReceiver, self).__init__()

#         # Calculate input dimensions based on y shape (128, 1, 1, 14, 76)
#         self.input_features = 14 * 76  # 1064

#         # Convolutional layers for feature extraction
#         self.conv_layers = nn.Sequential(
#             nn.Conv2d(1, 32, kernel_size=3, padding=1),
#             nn.BatchNorm2d(32),
#             nn.ReLU(),
#             nn.Conv2d(32, 64, kernel_size=3, padding=1),
#             nn.BatchNorm2d(64),
#             nn.ReLU(),
#             nn.Conv2d(64, 128, kernel_size=3, padding=1),
#             nn.BatchNorm2d(128),
#             nn.ReLU(),
#         )

#         # Calculate the size after convolutions
#         self.conv_output_size = 128 * 14 * 76  # Maintaining spatial dimensions

#         # Dense layers for signal processing
#         self.fc_layers = nn.Sequential(
#             nn.Linear(self.conv_output_size, 2048),
#             nn.ReLU(),
#             nn.Dropout(0.2),
#             nn.Linear(2048, 1024),
#             nn.ReLU(),
#             nn.Dropout(0.2),
#             nn.Linear(1024, 768),  # Output dimension matches required x shape
#         )

#     def forward(self, y, h_b, tau_b, h_out, no):
#         """
#         Forward pass of the neural receiver.

#         Args:
#             y (torch.Tensor): Received signal (128, 1, 1, 14, 76)
#             h_b (torch.Tensor): Channel impulse response (128, 1, 1, 1, 16, 10, 1)
#             tau_b (torch.Tensor): Delay profile (128, 1, 1, 10)
#             h_out (torch.Tensor): Channel response (128, 1, 1, 1, 16, 1, 76)
#             no (float): Noise power

#         Returns:
#             tuple: (x_hat, no_eff)
#                 - x_hat (torch.Tensor): Processed signal (128, 1, 1, 768)
#                 - no_eff (float): Effective noise power
#         """
#         batch_size = y.shape[0]
#         print("batch_size", batch_size)

#         # Reshape input for 2D convolution: (batch, channels, height, width)
#         y = y.view(batch_size, 1, 14, 76)
#         print("y shape", y.shape)

#         # Apply convolution layers
#         x = self.conv_layers(y)

#         # Flatten for dense layers
#         x = x.view(batch_size, -1)

#         # Apply dense layers
#         x = self.fc_layers(x)

#         # Reshape output to match required dimensions (128, 1, 1, 768)
#         x_hat = x.view(batch_size, 1, 1, -1)

#         # Calculate effective noise power
#         no_eff = self._calculate_effective_noise(x_hat, h_out, no)

#         return x_hat, no_eff

#     def _calculate_effective_noise(self, x_hat, h_out, no):
#         """
#         Calculate effective noise power based on the processed signal and channel response.

#         Args:
#             x_hat (torch.Tensor): Processed signal
#             h_out (torch.Tensor): Channel response
#             no (float): Original noise power

#         Returns:
#             float: Effective noise power
#         """
#         signal_power = torch.mean(torch.abs(x_hat) ** 2)
#         channel_effect = torch.mean(torch.abs(h_out) ** 2)
#         no_eff = no * signal_power * channel_effect
#         return float(no_eff.item())







In [145]:
import torch
import torch.nn as nn
import numpy as np


class NeuralReceiver(nn.Module):
    def __init__(self):
        """
        Neural Receiver that processes received signals to remove noise and channel effects.
        """
        super(NeuralReceiver, self).__init__()

        # Calculate input dimensions based on y shape (128, 1, 1, 14, 76)
        self.input_features = 14 * 76  # 1064

        # Convolutional layers for feature extraction
        self.conv_layers = nn.Sequential(
            nn.Conv2d(1, 32, kernel_size=3, padding=1),
            nn.BatchNorm2d(32),
            nn.ReLU(),
            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(),
            nn.Conv2d(64, 128, kernel_size=3, padding=1),
            nn.BatchNorm2d(128),
            nn.ReLU(),
        )

        # Calculate the size after convolutions
        self.conv_output_size = 128 * 14 * 76  # Maintaining spatial dimensions

        # Dense layers for signal processing
        self.fc_layers = nn.Sequential(
            nn.Linear(self.conv_output_size, 2048),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(2048, 1024),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(1024, 768),  # Output dimension matches required x shape
        )

    def forward(self, y, h_b, tau_b, h_out, no):
        """
        Forward pass of the neural receiver.

        Args:
            y (torch.Tensor): Received signal (128, 1, 1, 14, 76)
            h_b (torch.Tensor): Channel impulse response (128, 1, 1, 1, 16, 10, 1)
            tau_b (torch.Tensor): Delay profile (128, 1, 1, 10)
            h_out (torch.Tensor): Channel response (128, 1, 1, 1, 16, 1, 76)
            no (torch.Tensor): Noise power tensor of shape (1, 1)

        Returns:
            tuple: (x_hat, no_eff)
                - x_hat (torch.Tensor): Processed signal (128, 1, 1, 768)
                - no_eff (torch.Tensor): Effective noise power
        """
        batch_size = y.shape[0]
        # print("batch_size", batch_size)

        # Reshape input for 2D convolution: (batch, channels, height, width)
        y = y.view(batch_size, 1, 14, 76)
        # print("y shape", y.shape)

        # Apply convolution layers
        x = self.conv_layers(y)

        # Flatten for dense layers
        x = x.view(batch_size, -1)

        # Apply dense layers
        x = self.fc_layers(x)

        # Reshape output to match required dimensions (128, 1, 1, 768)
        x_hat = x.view(batch_size, 1, 1, -1)

        # Calculate effective noise power
        no_eff = self._calculate_effective_noise(x_hat, h_out, no)

        return x_hat, no_eff

    def _calculate_effective_noise(self, x_hat, h_out, no):
        """
        Calculate effective noise power based on the processed signal and channel response.

        Args:
            x_hat (torch.Tensor): Processed signal (batch_size, 1, 1, 768)
            h_out (torch.Tensor): Channel response (batch_size, 1, 1, 1, 16, 1, 76)
            no (torch.Tensor): Original noise power tensor of shape (1, 1)

        Returns:
            torch.Tensor: Effective noise power
        """
        # Calculate signal power while preserving batch dimension
        # Calculate signal power while preserving batch dimension
        signal_power = torch.mean(torch.abs(x_hat) ** 2)  # scalar

        # Calculate channel effect
        channel_effect = torch.mean(torch.abs(h_out) ** 2)  # scalar

        # Calculate effective noise power
        no_eff = no * signal_power * channel_effect

        # Ensure shape is (1, 1)
        no_eff = no_eff.view(1, 1)

        return no_eff  # Returns tensor of shape (1, 1)

In [83]:
def numpy_to_torch(arr):
    """Convert numpy array to torch tensor."""
    return torch.from_numpy(arr).float()

In [160]:
print("y shape", y.shape)
print("x shape", x.shape)
print("no shape", no)
print("h_b shape", h_b.shape)
print("tau_b shape", tau_b.shape)
print("h_out shape", h_out.shape)

y shape torch.Size([128, 1, 1, 14, 76])
x shape torch.Size([128, 1, 1, 768])
no shape tensor([[0.0158]])
h_b shape torch.Size([128, 1, 1, 1, 16, 10, 1])
tau_b shape torch.Size([128, 1, 1, 10])
h_out shape torch.Size([128, 1, 1, 1, 16, 1, 76])


In [95]:
# Convert numpy arrays to torch tensors

y_torch = torch.from_numpy(y).float().requires_grad_(True)
h_b_torch = torch.from_numpy(h_b).float().requires_grad_(True)
tau_b_torch = torch.from_numpy(tau_b).float().requires_grad_(True)
h_out_torch = torch.from_numpy(h_out).float().requires_grad_(True)

no_tensor = torch.full((1, 1), no, dtype=torch.float32)


print("y_torch shape", y_torch.shape)
print("h_b_torch shape", h_b_torch.shape)
print("tau_b_torch shape", tau_b_torch.shape)
print("h_out_torch shape", h_out_torch.shape)
print("no_tensor shape", no_tensor)


y_torch shape torch.Size([128, 1, 1, 14, 76])
h_b_torch shape torch.Size([128, 1, 1, 1, 16, 10, 1])
tau_b_torch shape torch.Size([128, 1, 1, 10])
h_out_torch shape torch.Size([128, 1, 1, 1, 16, 1, 76])
no_tensor shape tensor([[0.0158]])


In [107]:
# Example usage
# x_hat2, no_eff2 = process_signal(y, x, no, h_b, tau_b, h_out)
receiver = NeuralReceiver()
x_hat2, no_eff2 = receiver(y_torch, h_b_torch, tau_b_torch, h_out_torch, no_tensor)

# Convert output back to numpy
x_hat_np = x_hat.detach().numpy()
float_value = no_eff2.item()

In [110]:
print(x_hat_np.shape)
print(float_value)

(128, 1, 1, 768)
6.14891032455489e-05


# Evaluate neural receiver

In [113]:
b_hat2, llr_est = eval_transceiver.demapper_decision(x_hat=x_hat_np, no_eff=float_value)

In [114]:
from deepMIMO5 import calculate_BER 
BER = calculate_BER(b, b_hat2)
print("BER Value:", BER)

BER Value: 0.5009053548177084


In [26]:
# # Add this line to display the figures
# import matplotlib.pyplot as plt

# plt.show()  #

# Plotting the testing of neural receiver,
- Testing on list of enobdb
- Testing on different bits

result:
- without training, It maintains average ber rate of 0.5

In [16]:
# Define the range of Eb/No values
ebno_db_range = range(1, 11)  # Eb/No values from 1 to 10

# Store results
results = []


for ebno_db in ebno_db_range:
    # Generate bits

    # Call the gettransmitSignal method
    y, x_rg, x, b, no, h_b, tau_b, h_out = eval_transceiver.gettransmitSignal(
        b=b,
        ebno_db=ebno_db,  # Current Eb/No value
        perfect_csi=False,
    )

    # Process the received signal using the neural receiver
    # Example usage
    x_hat2, no_eff2 = process_signal(y, x, no, h_b, tau_b, h_out)

    # Calculate Bit Error Rate (BER)
    # Assuming b and x_hat are in the same format, you may need to convert x_hat to bits
    # For simplicity, let's assume x_hat is already in the correct format
    # b_hat = np.round(x_hat)  # Example: rounding to get estimated bits
    # ber = np.mean(b != b_hat)  # Calculate BER

    # # Store the results
    # results.append((ebno_db, ber))
    b_hat2, llr_est = eval_transceiver.demapper_decision(x_hat=x_hat2, no_eff=no_eff2)
    BER = calculate_BER(b, b_hat2)
    results.append((ebno_db, BER))

# Print the results
for ebno_db, ber in results:
    print(f"Eb/No (dB): {ebno_db}, BER: {ber:.4f}")

h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)


d:\Research\AIsensing\deeplearning\AIsim_main2.py:1207: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  plt.figure()


h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
Eb/No (dB): 1, BER: 0.5023
Eb/No (dB): 2, BER: 0.5003
Eb/No (dB): 3, BER: 0.5005
Eb/No (dB): 4, BER: 0.5001
Eb/No (dB): 5, BER: 0.5018
Eb/No (dB): 6, BER: 0.5023
Eb/No (dB): 7, BER: 0.5017
Eb/No (dB): 8, BER: 0.5013
Eb/No (dB): 9, BER: 0.4993
Eb/No (dB): 10, BER: 0.5008


In [21]:
# Unzip the results for plotting
ebno_values, ber_values = zip(*results)

import matplotlib.pyplot as plt
# Plotting
# Assuming you have already generated the plot
plt.figure(figsize=(10, 6))
plt.semilogy(ebno_values, ber_values, marker="o", linestyle="-", color="b")
plt.title("BER vs Eb/No")
plt.xlabel("Eb/No (dB)")
plt.ylabel("Bit Error Rate (BER)")
plt.grid(True, which="both", linestyle="--", linewidth=0.5)
plt.xticks(ebno_values)  # Set x-ticks to be the Eb/No values
plt.ylim(1e-6, 1)  # Set y-limits for better visibility
plt.xlim(min(ebno_values), max(ebno_values))

# Define the path where you want to save the PDF
pdf_path = r"D:\Research\AIsensing\deeplearning\data\ber_vs_ebno.pdf"

# Save the plot as a PDF file in the specified directory
plt.savefig(pdf_path, format="pdf")

# Optionally, you can close the plot to free up memory
plt.close()

In [22]:
import os

os.startfile("ber_vs_ebno.pdf")

In [19]:
# Define the number of different bit sequences to test
num_bit_sequences = 5

# Store results
results = []

for i in range(num_bit_sequences):
    # Generate bits
    binary_source = BinarySource()
    NUM_STREAMS_PER_TX = eval_transceiver.num_streams_per_tx
    k = eval_transceiver.k
    b = binary_source(
        [BATCH_SIZE, 1, NUM_STREAMS_PER_TX, k]
    )  # [batch_size, num_tx, num_streams_per_tx, num_databits]

    # Call the gettransmitSignal method
    y, x_rg, x, b, no, h_b, tau_b, h_out = eval_transceiver.gettransmitSignal(
        b=b,
        ebno_db=15.0,  # Use a fixed Eb/No value for testing
        perfect_csi=False,
    )

    # Process the received signal using the neural receiver
    x_hat2, no_eff2 = process_signal(y, x, no, h_b, tau_b, h_out)

    # Calculate Bit Error Rate (BER)
    b_hat2, llr_est = eval_transceiver.demapper_decision(x_hat=x_hat2, no_eff=no_eff2)
    BER = calculate_BER(b, b_hat2)

    # Store the results
    results.append((i + 1, BER))  # Store the sequence number and BER

# Print the results
for seq_num, ber in results:
    print(f"Bit Sequence {seq_num}: BER: {ber:.4f}")

h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
Bit Sequence 1: BER: 0.5000
Bit Sequence 2: BER: 0.5002
Bit Sequence 3: BER: 0.4995
Bit Sequence 4: BER: 0.5006
Bit Sequence 5: BER: 0.5007


In [ ]:
# bers, blers, BERs = sim_bersingle2(
#     channeldataset="deepmimo",
#     channeltype="ofdm",
#     NUM_BITS_PER_SYMBOL=2,
#     EBN0_DB_MIN=-5.0,
#     EBN0_DB_MAX=25.0,
#     BATCH_SIZE=128,
#     NUM_UT=1,
#     NUM_BS=1,
#     NUM_UT_ANT=1,
#     NUM_BS_ANT=16,
#     showfigure=True,  # or whatever value you want
#     datapathbase="data/",
#     dataset_folder=dataset_folder
# )

# Training phase

y shape (128, 1, 1, 14, 76)
x shape (128, 1, 1, 768)
no shape 0.015811387
h_b shape (128, 1, 1, 1, 16, 10, 1)
tau_b shape (128, 1, 1, 10)
h_out shape (128, 1, 1, 1, 16, 1, 76)


In [129]:
# Create a dataset to store b, y, x, tau_b, and h_out
dataset = []

# Number of samples you want to generate
num_samples = 10  # Adjust as needed

for _ in range(num_samples):
    # Generate bits
    binary_source = BinarySource()
    NUM_STREAMS_PER_TX = eval_transceiver.num_streams_per_tx
    k = eval_transceiver.k
    b = binary_source(
        [BATCH_SIZE, 1, NUM_STREAMS_PER_TX, k]
    )  # [batch_size, num_tx, num_streams_per_tx, num_databits]

    # Call the gettransmitSignal method
    y, x_rg, x, b, no, h_b, tau_b, h_out = eval_transceiver.gettransmitSignal(
        b=b,  # Pass the generated bits
        ebno_db=15.0,  # Example Eb/No value
        perfect_csi=False,  # Example for perfect CSI
    )
    print("type(y)", type(y))
    print("type(x_rg)", type(x_rg))
    print("type(x)", type(x))
    print("no", no)
    # Store the generated data in the dataset, including tau_b and h_out
    dataset.append((y, x, b, no, h_b, tau_b, h_out))
    # y, x_rg, x, b, no, h_b, tau_b, h_out

# Now dataset contains tuples of (b, y, x, tau_b, h_out) for training

h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class 'numpy.ndarray'>
type(x_rg) <class 'numpy.ndarray'>
type(x) <class 'numpy.ndarray'>
no 0.015811387
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class 'numpy.ndarray'>
type(x_rg) <class 'numpy.ndarray'>
type(x) <class 'numpy.ndarray'>
no 0.015811387
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class 'numpy.ndarray'>
type(x_rg) <class 'numpy.ndarray'>
type(x) <class 'numpy.ndarray'>
no 0.015811387
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class 'numpy.ndarray'>
type(x_rg) <class 'numpy.ndarray'>
type(x) <class 'numpy.ndarray'>
no 0.015811387
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class 'numpy.ndarray'>
type(x_rg) <class 'numpy.ndarray'>
type(x) <class 'numpy.ndarray'>
no 0.015811387
h_freq shape: (128, 1, 1, 1, 16, 1, 76)
y shape: (128, 1, 1, 14, 76)
type(y) <class '

In [130]:
# Print the dataset
for index, (y, x, b, no, h_b, tau_b, h_out) in enumerate(dataset):
    print(f"Sample {index + 1}:")
    print("Received Signal (y):", y.shape)
    print("Transmitted Signal (x):", x.shape)
    print("Bits (b):", b.shape)
    print("Noise Power (no):", no)  # Assuming no is a float, print its value
    print("Channel Impulse Response (h_b):", h_b.shape)
    print("Delay Profile (tau_b):", tau_b.shape)
    print("Channel Response (h_out):", h_out.shape)
    print("-" * 40)  # Separator for readability

Sample 1:
Received Signal (y): (128, 1, 1, 14, 76)
Transmitted Signal (x): (128, 1, 1, 768)
Bits (b): (128, 1, 1, 1536)
Noise Power (no): 0.015811387
Channel Impulse Response (h_b): (128, 1, 1, 1, 16, 10, 1)
Delay Profile (tau_b): (128, 1, 1, 10)
Channel Response (h_out): (128, 1, 1, 1, 16, 1, 76)
----------------------------------------
Sample 2:
Received Signal (y): (128, 1, 1, 14, 76)
Transmitted Signal (x): (128, 1, 1, 768)
Bits (b): (128, 1, 1, 1536)
Noise Power (no): 0.015811387
Channel Impulse Response (h_b): (128, 1, 1, 1, 16, 10, 1)
Delay Profile (tau_b): (128, 1, 1, 10)
Channel Response (h_out): (128, 1, 1, 1, 16, 1, 76)
----------------------------------------
Sample 3:
Received Signal (y): (128, 1, 1, 14, 76)
Transmitted Signal (x): (128, 1, 1, 768)
Bits (b): (128, 1, 1, 1536)
Noise Power (no): 0.015811387
Channel Impulse Response (h_b): (128, 1, 1, 1, 16, 10, 1)
Delay Profile (tau_b): (128, 1, 1, 10)
Channel Response (h_out): (128, 1, 1, 1, 16, 1, 76)
---------------------

In [141]:
import torch
from torch.utils.data import DataLoader, TensorDataset

# Prepare the dataset
def prepare_data(dataset):
  tensor_dataset = []
  
  for index, (y, x, b, no, h_b, tau_b, h_out) in enumerate(dataset):
    # Convert numpy arrays to torch tensors
    y_torch = torch.from_numpy(y).float().requires_grad_(True)
    x_torch = torch.from_numpy(x).float().requires_grad_(True)
    b_torch = torch.from_numpy(b).float().requires_grad_(True)
    no_torch = torch.full((1, 1), no, dtype=torch.float32)
    h_b_torch = torch.from_numpy(h_b).float().requires_grad_(True)
    tau_b_torch = torch.from_numpy(tau_b).float().requires_grad_(True)
    h_out_torch = torch.from_numpy(h_out).float().requires_grad_(True)
    tensor_dataset.append((y_torch, x_torch, b_torch, no_torch, h_b_torch, tau_b_torch, h_out_torch))

  return tensor_dataset  

tensor_dataset = prepare_data(dataset)    
   
# dataloader = DataLoader(
#     dataset,
#     batch_size=1,
#     shuffle=True,
# )   

In [147]:
import torch
import torch.optim as optim

# Initialize the model
model = NeuralReceiver()

# Define loss function and optimizer
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Number of epochs
num_epochs = 10

# Training loop
for epoch in range(num_epochs):
    total_loss = 0
    for y, x, b, no, h_b, tau_b, h_out in tensor_dataset:
        # Forward pass
        x_hat, no_eff = model(y, h_b, tau_b, h_out, no)

        # Compute loss
        loss = criterion(x_hat, x)

        # Backward pass and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    # Print average loss for the epoch
    avg_loss = total_loss / len(tensor_dataset)
    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {avg_loss:.4f}")

Epoch [1/10], Loss: 3.5235
Epoch [2/10], Loss: 0.5003
Epoch [3/10], Loss: 0.5004
Epoch [4/10], Loss: 0.5004
Epoch [5/10], Loss: 0.5003
Epoch [6/10], Loss: 0.5002
Epoch [7/10], Loss: 0.5002
Epoch [8/10], Loss: 0.5002
Epoch [9/10], Loss: 0.5001
Epoch [10/10], Loss: 0.5001


In [171]:
# Evaluation
model.eval()
with torch.no_grad():
    total_ber = 0
    num_samples = 0
    for y, x, b, no, h_b, tau_b, h_out in tensor_dataset:
        x_hat, no_eff = model(y, h_b, tau_b, h_out, no)

        # Use the provided evaluation functions
        b_hat, _ = eval_transceiver.demapper_decision(
            x_hat=x_hat.numpy(), no_eff=no_eff.item()
        )
        ber = calculate_BER(b.numpy(), b_hat)

        total_ber += ber * len(y)
        num_samples += len(y)

    average_ber = total_ber / num_samples
    print(f"Average BER: {average_ber:.6f}")

Average BER: 0.500286


In [150]:
import os
# Define the model save path
model_path = "./TrainedModel/neural_receiver_model.pth"

# Create the directory if it doesn't exist
os.makedirs(os.path.dirname(model_path), exist_ok=True)

# Save the model
torch.save(model.state_dict(), model_path)
print(f"Model saved to {model_path}")

Model saved to ./TrainedModel/neural_receiver_model.pth


# Training track: 2

In [172]:
# Initialize model, loss function, and optimizer
model = NeuralReceiver()
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Set the target BER
target_ber = 0.1

# Training loop
epoch = 0
max_epochs = 1000  # Set a maximum number of epochs to prevent infinite loops
while True:
    epoch += 1
    total_loss = 0

    for y, x, b, no, h_b, tau_b, h_out in tensor_dataset:
        # Forward pass
        x_hat, no_eff = model(y, h_b, tau_b, h_out, no)

        # Compute loss
        loss = criterion(x_hat, x)

        # Backward pass and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    # Print average loss for the epoch
    avg_loss = total_loss / len(tensor_dataset)
    print(f"Epoch [{epoch}/{max_epochs}], Loss: {avg_loss:.4f}")

    # Evaluate BER
    model.eval()
    with torch.no_grad():
        total_ber = 0
        num_samples = 0
        for y, x, b, no, h_b, tau_b, h_out in tensor_dataset:
            x_hat, no_eff = model(y, h_b, tau_b, h_out, no)

            # Convert to numpy for evaluation
            x_hat_np = x_hat.numpy()
            no_eff_np = no_eff.item()

            # Use the provided evaluation functions
            b_hat, _ = eval_transceiver.demapper_decision(
                x_hat=x_hat_np, no_eff=no_eff_np
            )
            ber = calculate_BER(b.numpy(), b_hat)

            total_ber += ber * len(y)
            num_samples += len(y)

        average_ber = total_ber / num_samples
        print(f"Average BER: {average_ber:.6f}")

    # Check if target BER is achieved
    if average_ber < target_ber:
        print(f"Target BER of {target_ber} achieved. Training complete.")
        break

    # Check if maximum epochs reached
    if epoch >= max_epochs:
        print(f"Maximum epochs ({max_epochs}) reached. Training stopped.")
        break

    model.train()  # Set model back to training mode

print(f"Final Average BER: {average_ber:.6f}")

Epoch [1/1000], Loss: 3.1942
Average BER: 0.500741
Epoch [2/1000], Loss: 0.5004
Average BER: 0.500759
